### PR TITLE
policy: Port Range Match Label Fix

### PR DESCRIPTION
**Pull request by nathanjsweet**
The original port range changes for EgressCoversContext
inadvertantly undid implicit logic underlying
using the go builtin map. That is that deny entries
would always be correctly identified by label, based
on the fact that no more than one port-protocol key
could match a given set of labels. After the introduction
of port ranges this implicit assumption was false,
but no logic to prioritize deny entries over allows
was introduced. Instead the code naively looked up
the longest prefix match rule by port-protocol.
This works most of the time, but fails for deny
entries of lesser port-protocol prefix matches
when a longer prefix allow is present.
    
This commit introduces a new method on the
L4PolicyMap `MatchesLabels` that looks up applicable
rules by label using the longest prefix match
lookup by port-protocol, but correctly finds
deny entries that may have lower port-protocol
precedence.

```release-note
policy: Fix policy cache covers context lookup.
```

### CODE REVIEW

File: pkg/policy/distillery_test.go [hunk 1]
```diff
@@ -1831,3 +1836,203 @@ func Test_EnsureEntitiesSelectableByCIDR(t *testing.T) {
...
+				{1, 2, true},
+				{79, 0, true},
+				{80, 0, false},
+				{81, 0, true},
+				{1023, 1024, true},
+				{1025, 1026, false},
+			},
+		},
+		{
+			name: "wildcard deny (*) covers broad allow range (1-1024)",
```
Comment (doniacld) [inline pkg/policy/distillery_test.go:1907]: I do not see the difference with the case below 🧐

**Review by doniacld — APPROVED** [review]
Only one comment! Otherwise, lgtm! :)

File: pkg/policy/distillery_test.go (hunk 1 above)
Comment (nathanjsweet) [reply pkg/policy/distillery_test.go:1907]: Good catch.

**Comment by gandro** [comment]
Added backport author as the code does not compile on v1.16

**Pull request merged**